### PR TITLE
Uplift CAPI to v1.3.0 minor release

### LINKED ISSUE #2769: Adopt and test CAPI v1.3 with beta release
This is an issue to track updating CAPI and the CAPI test framework to v1.3.0-beta.0 now that it is available.

https://kubernetes.slack.com/archives/C8TSNPY4T/p1667414755279279

Release: https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.3.0-beta.0
Migration guide: https://github.com/kubernetes-sigs/cluster-api/blob/v1.3.0-beta.0/docs/book/src/developer/providers/v1.2-to-v1.3.md

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR attempts to adopt and test CAPI **v1.3.0** release in CAPZ, more specifically:

By following https://github.com/kubernetes-sigs/cluster-api/blob/v1.3.0/docs/book/src/developer/providers/v1.2-to-v1.3.md, this PR uplifts dependencies:
- ~bumps Go version to 1.19.X~ (taken care by separate [PR](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2775)) ✅
- k8s.io/* to v0.25.0
- sigs.k8s.io/controller-tools to v0.10.0
- sigs.k8s.io/controller-runtime to v0.13.1
- Ginkgo to v2 (2.5.0) and Gomega to v1.24.1 and related changes
- cert-manager to v1.10.0
- kind version to v0.17.0
- github.com/joelanford/go-apidiff to 0.5.0
- KUBERNETES_VERSION to v1.24.6
- ETCD_VERSION_UPGRADE_TO 3.5.4-0
- COREDNS_VERSION_UPGRADE_TO v1.9.3
- KUBERNETES_VERSION_UPGRADE_TO 1.25
- KUBERNETES_VERSION_UPGRADE_FROM 1.24

Note: `make verify` & `make verify-conversions` and `make verify-modules` targets have been run and are passing locally. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2769 

**Special notes for your reviewer**:
~Do we have to bump also images in https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes-sigs/cluster-api-provider-azure to use the `v1.25` image with Go `1.19`?~ (taken care by separate [PR](https://github.com/kubernetes/test-infra/pull/27894)) ✅ 

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
> Do we have to bump also images in https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes-sigs/cluster-api-provider-azure to use the `v1.25` image with Go `1.19`?

/cc @CecileRobertMichon @jackfrancis hi folks, can you take a look and suggest my assumption is correct? All of the CI jobs are complaining about

**Comment by CecileRobertMichon** [comment]
> Do we have to bump also images in https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes-sigs/cluster-api-provider-azure to use the v1.25 image with Go 1.19?

Yes I believe so

Thanks for doing this work!

**Comment by furkatgofurov7** [comment]
> 
> Yes I believe so

@CecileRobertMichon here it is: https://github.com/kubernetes/test-infra/pull/27894

**Comment by furkatgofurov7** [comment]
/hold

I will separate Go version update changes from this PR to have a smoother transition overall 🤞🏼

**Comment by furkatgofurov7** [comment]
> I will separate Go version update changes from this PR to have a smoother transition overall 🤞🏼

xref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2775
I will rebase on top of #2775 once it lands

**Comment by jackfrancis** [comment]
@furkatgofurov7 go 1.19 PR landed, are you able to rebase on top of that?

Thank you!

**Comment by furkatgofurov7** [comment]
> * `e2e/e2e-exp/upgrade `(they mostly complain about incompatible clusterctl version) which does not make sense at least considering my changes in place..

@CecileRobertMichon @mboersma any suggestions on how to debug these tests?

File: api/v1alpha3/azurecluster_types.go
Comment (mboersma) [inline api/v1alpha3/azurecluster_types.go:67]: When modifying these URLs, consider removing the `/en-us` part. It's a more readable URL and will redirect to a version localized to your browser's language (if available).

https://learn.microsoft.com/azure/reliability/availability-zones-overview

**Comment by mboersma** [comment]
For _2) ci-entrypoint_, It seems like somehow the `Byf` func is being called from a context other than Ginkgo. The script doesn't call any tests, it just sets up a cluster, and I'm having trouble seeing what might have called this code outside of e2e tests.

File: api/v1alpha3/azurecluster_types.go
Comment (furkatgofurov7) [reply api/v1alpha3/azurecluster_types.go]: @mboersma thanks, I appended the fix to the old doc lint fixing [commit](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2773/commits/353a39413d8be27d0bba1cb8aa4a383fc806dc13) and should be fine now

**Comment by furkatgofurov7** [comment]
> For _2) ci-entrypoint_, It seems like somehow the `Byf` func is being called from a context other than Ginkgo. The script doesn't call any tests, it just sets up a cluster, and I'm having trouble seeing what might have called this code outside of e2e tests.

That is very strange indeed. I found https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_clusterproxy.go#L91 being called from https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/logger.go#L74

**Comment by mboersma** [comment]
That agrees with the stack trace I got from adding `debug.PrintStack` to the `Byf` func:

```golang
  INFO: Creating log watcher for controller capz-system/capz-controller-manager, pod capz-controller-manager-6564c4dc85-z5zxg, container manager
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x1c
sigs.k8s.io/cluster-api-provider-azure/test/e2e.Byf({0x105cac87c, 0x23}, {0x14000fffd78, 0x2, 0x2})
	/Users/matt/projects/cluster-api-provider-azure/test/e2e/common.go:85 +0x30
sigs.k8s.io/cluster-api-provider-azure/test/e2e.(*AzureClusterProxy).CollectWorkloadClusterLogs(0x14000542300, {0x1065e4bb0, 0x140001a2008}, {0x16b8aaf5b, 0x7}, {0x16b8aaf43, 0xb}, {0x140000525a0, 0x4f})
	/Users/matt/projects/cluster-api-provider-azure/test/e2e/azure_clusterproxy.go:92 +0x9c
main.main()
	/Users/matt/projects/cluster-api-provider-azure/test/logger.go:74 +0x4dc
Ginkgo detected an issue with your spec structure
By(fmt.Sprintf(format, a...))
/Users/matt/projects/cluster-api-provider-azure/test/e2e/common.go:86
  It looks like you are calling By outside of a running spec.  Make sure you
  call By inside a runnable node such as It or BeforeEach and not inside the
  body of a container such as Describe or Context.

  Learn more at: http://onsi.github.io/ginkgo/#documenting-complex-specs-by

exit status 1
```

**Comment by mboersma** [comment]
We do call `CollectWorkloadClusterLogs` from the `main` func in logger.go. This isn't a context where we would expect to be inside an `It` or similar Ginkgo container.

It's being called from here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/scripts/ci-entrypoint.sh#L174

Maybe disallowing `By` here is new v2 behavior? I tried registering Ginkgo's `Fail` func but got the same error.

I guess we could work around it by using logging or `fmt.Printf` in `CollectWorkloadClusterLogs` instead of `By`.

**Comment by furkatgofurov7** [comment]
> We do call `CollectWorkloadClusterLogs` from the `main` func in logger.go. This isn't a context where we would expect to be inside an `It` or similar Ginkgo container.
> 
> It's being called from here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/scripts/ci-entrypoint.sh#L174
> 
> Maybe disallowing `By` here is new v2 behavior? I tried registering Ginkgo's `Fail` func but got the same error.
> 
> I guess we could work around it by using logging or `fmt.Printf` in `CollectWorkloadClusterLogs` instead of `Byf`.

Makes sense, even though I have not found the reference to that in Ginkgo v2, that was the only thing I spotted out in my prev. comment. I replaced now `Byf` with `Logf` to make it more aligned with e2e tests files.

File: Tiltfile
Comment (furkatgofurov7) [inline Tiltfile:23]: note to myself: update cert-manager version to v1.10.0

**Comment by furkatgofurov7** [comment]
> We do call `CollectWorkloadClusterLogs` from the `main` func in logger.go. This isn't a context where we would expect to be inside an `It` or similar Ginkgo container.
> 
> It's being called from here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/scripts/ci-entrypoint.sh#L174
> 
> Maybe disallowing `By` here is new v2 behavior? I tried registering Ginkgo's `Fail` func but got the same error.
> 
> I guess we could work around it by using logging or `fmt.Printf` in `CollectWorkloadClusterLogs` instead of `Byf`.

@mboersma I had to change  `CollectWorkloadClusterLogs` and also other functions called within it and now ci-entrypoint job seems to be happy

**Comment by mboersma** [comment]
I ran `GINKGO_FOCUS="Creating a highly available cluster" SKIP_CLEANUP=true ./scripts/ci-e2e.sh` locally on this branch and the workload cluster it created appears fine and passed its tests. It's just that something goes wrong in the `AfterEach` at line 115 of azure_test.go, and I'm having trouble deciphering this new stack trace output.

**Comment by mboersma** [comment]
BTW @furkatgofurov7 we should remove the outdated Ginkgo flags from line 665 of the Makefile:

```
LOCAL_GINKGO_ARGS ?= -stream --progress
```

In fact, we can probably get rid of `LOCAL_GINKGO_ARGS` altogether since its only purpose was to set these two flags, which are no longer supported.

**Comment by furkatgofurov7** [comment]
> To see what's up with `pull-cluster-api-provider-azure-e2e`, I ran `GINKGO_FOCUS="Creating a highly available cluster" SKIP_CLEANUP=true ./scripts/ci-e2e.sh` locally on this branch and the workload cluster it created appears fine and passed its tests. It's just that something goes wrong in the `AfterEach` at line 115 of azure_test.go, and I'm having trouble deciphering this new stack trace output.

@mboersma thanks a lot, that was the case for me as well. Besides, after removing the `LOCAL_GINKGO_ARGS` e2e-azure has passed for the first time now however e2e-exp failed again. I think the failure is more or less similar to e2e-azure test in e2e-exp (l115 in azure_test.go) and I'll try my luck to retriggering it and see if it helps. 

/test pull-cluster-api-provider-azure-e2e-exp

**Comment by furkatgofurov7** [comment]
@mboersma  so, re-triggering did not help and this is the full stack trace from e2e-exp:

```
{A suite timeout occurred timedout [TIMEDOUT] A suite timeout occurred
In [AfterEach] at: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:115 @ 11/28/22 23:29:25.228

This is the Progress Report generated when the suite timeout occurred:
  Workload cluster creation Creating an AKS cluster [EXPERIMENTAL][Managed Kubernetes] with a single control plane node and 1 node (Spec Runtime: 55m2.549s)
    /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:557
    In [AfterEach] (Node Runtime: 24m12.554s)
      /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:115
      At [By Step] Waiting for cluster capz-e2e-8f16at-aks to be deleted (Step Runtime: 4m18.307s)
        /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.0-rc.1/framework/ginkgoextensions/output.go:35

      Spec Goroutine
      goroutine 43554 [select]
        github.com/onsi/gomega/internal.(*AsyncAssertion).match(0xc00023db90, {0x4010478?, 0x5884290}, 0x1, {0x0, 0x0, 0x0})
          /home/prow/go/pkg/mod/github.com/onsi/gomega@v1.24.1/internal/async_assertion.go:426
        github.com/onsi/gomega/internal.(*AsyncAssertion).Should(0xc00023db90, {0x4010478, 0x5884290}, {0x0, 0x0, 0x0})
          /home/prow/go/pkg/mod/github.com/onsi/gomega@v1.24.1/internal/async_assertion.go:110
        sigs.k8s.io/cluster-api/test/framework.WaitForClusterDeleted({0x4020260?, 0xc00005a0d0}, {{0x7fa3a828d290?, 0xc000f1d730?}, 0xc000b7d100?}, {0xc002bbfe00, 0x2, 0x2})
          /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.0-rc.1/framework/cluster_helpers.go:176
        sigs.k8s.io/cluster-api/test/framework.DeleteAllClustersAndWait({0x4020260?, 0xc00005a0d0}, {{0x4030758?, 0xc000f1d730?}, {0xc000f6d5f0?, 0x1?}}, {0xc002bbfe00, 0x2, 0x2})
          /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.0-rc.1/framework/cluster_helpers.go:272
      > sigs.k8s.io/cluster-api-provider-azure/test/e2e.dumpSpecResourcesAndCleanup({0x4020260, 0xc00005a0d0}, {{0x3ba9485, 0x17}, {0x4032650, 0xc00126b3d0}, {0xc0005c6000, 0xf}, 0xc000921a20, 0xc0001d2bb0, ...})
          /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/common.go:170
            | 	deleteTimeoutConfig = "wait-delete-cluster-aks"
            | }
            > framework.DeleteAllClustersAndWait(ctx, framework.DeleteAllClustersAndWaitInput{
            | 	Client:    input.ClusterProxy.GetClient(),
            | 	Namespace: input.Namespace.Name,
      > sigs.k8s.io/cluster-api-provider-azure/test/e2e.glob..func1.2()
          /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:136
            | 	SkipResourceGroupCheck: skipResourceGroupCheck,
            | }
            > dumpSpecResourcesAndCleanup(ctx, cleanInput)
            | Expect(os.Unsetenv(AzureResourceGroup)).To(Succeed())
            | Expect(os.Unsetenv(AzureVNetName)).To(Succeed())
        github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3({0x1bb280e, 0xc002a87680})
          /home/prow/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.5.0/internal/node.go:445
        github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
          /home/prow/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.5.0/internal/suite.go:820
        github.com/onsi/ginkgo/v2/internal.(*Suite).runNode
          /home/prow/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.5.0/internal/suite.go:807
}
```
which is not very informative. I'll try to debug it more, but wanted to ask if there is any easy way to get access to the live cluster when tests are running?

**Comment by furkatgofurov7** [comment]
azure-e2e tests were passing yesterday, however they again started to fail, which makes me think there is something going wrong unrelated to my changes.. 

/test pull-cluster-api-provider-azure-e2e
/test pull-cluster-api-provider-azure-e2e-exp

**Comment by CecileRobertMichon** [comment]
@furkatgofurov7 do you need any help on this?

/milestone v1.7

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 do you need any help on this?

@CecileRobertMichon that would be great! The best I could achieve so far was having 2 failing tests (`apidiff`, `e2e-exp`), but after the rebase e2e started failing again :(.  The main challenge to fix was, I do not enough how you guys debug your failing tests (access to live env etc), and the logs from the runs seem to be super unhelpful 😀 . Also, please let me know and I am absolutely fine, if this is a blocker and working on your own rather than helping to debug it is easier,  I will close it.

**Comment by mboersma** [comment]
/retest pull-cluster-api-provider-azure-e2e

GitHub API rate limit exceeded. I haven't actually seen this before in e2e. Maybe we need to set `GITHUB_TOKEN` for this job?

File: Makefile
Comment (Jont828) [inline Makefile:670]: Just wondering, is this no longer needed in Ginkgo v2?

**Review by Jont828 — COMMENTED** [review]
Thanks for this contribution! I like that we're using `Logf` now too. Overall, this looks fine, I think we just need to debug the e2e tests a little more.

File: test/e2e/config/azure-dev.yaml
Comment (CecileRobertMichon) [inline test/e2e/config/azure-dev.yaml:50]: since the actual v1.3.0 release dropped today, should we skip the rc and directly bump to v1.3.0?

**Comment by CecileRobertMichon** [comment]
Regarding failing e2e tests, it looks like the overall suite is timing out
```
  [38;5;214m[TIMEDOUT] A suite timeout occurred[0m
```

I think there may be an issue with Ginkgo parallelism, see https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2773/pull-cluster-api-provider-azure-e2e/1598381939977359360/build-log.txt

I also see a bunch of `INFO: "" started ` (unexpected empty string for the spec name)

And some deprecated functionality warnings at the bottom.

Looks like Ginkgo v2 added some timestamps in a bunch of places which makes our homemade timestamp printing redundant but that's more cosmetic, we can clean that up later as a follow up.

File: Makefile
Comment (furkatgofurov7) [reply Makefile:670]: @Jont828 that was removed based on this https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2773#issuecomment-1329727450 above

File: test/e2e/config/azure-dev.yaml
Comment (furkatgofurov7) [reply test/e2e/config/azure-dev.yaml]: Done

**Comment by mboersma** [comment]
I have a fix for the conformance test warnings, but that's just a deprecation warning, not an error. I'm looking at the AKS failure in e2e-exp.

**Comment by mboersma** [comment]
`GINKGO_SKIP="" GINKGO_FOCUS="Creating an AKS cluster" ./scripts/ci-e2e.sh` passed for me locally, so I didn't reproduce what we've seen several times in the pull-cluster-api-provider-azure-e2e-exp job. Maybe I got lucky with the cleanup, which seems to be an issue.

Maybe we aren't accounting for the new [timeout behavior](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#timeout-behavior) and AKS just takes longer than 1h?

**Comment by mboersma** [comment]
Also the Gingko v2 migration docs say:

> Finally, the default timeout has been reduced from 24h down to 1h

In the Makefile, we don't specify a timeout. So probably before, the tests defaulted to an overly generous timeout value (24h), and now it's just 1h. Maybe we should add `--timeout=3h` (or 24h?) to `GINKGO_ARGS` default in the Makefile?

**Comment by furkatgofurov7** [comment]
> Also the Gingko v2 migration docs say:
> 
> > Finally, the default timeout has been reduced from 24h down to 1h
> 
> In the Makefile, we don't specify a timeout. So probably before, the tests defaulted to an overly generous timeout value (24h), and now it's just 1h. Maybe we should add `--timeout=3h` (or 24h?) to `GINKGO_ARGS` default in the Makefile?

@mboersma great find, that should well be the case indeed. I will update the PR shortly to add the timeout

File: Makefile
Comment (mboersma) [inline Makefile:670]: @furkatgofurov7 I was wrong--sorry! I thought `--progress` had been removed but it's still there in Ginkgo v2. Could you restore this with just the `--progress` flag? 🎖️

**Comment by CecileRobertMichon** [comment]
looks like a delete timeout on the last one...


/retest

File: Makefile
Comment (furkatgofurov7) [reply Makefile]: Right, I reverted it back and left `--progress` only in the args

Comment (furkatgofurov7) [reply Makefile]: @mboersma actually that looks like has been removed according to prow logs:
```
 Test Suite Failed
You're using deprecated Ginkgo functionality:
=============================================
  --progress is deprecated .  The funcitonality provided by --progress was confusing and is no longer needed.  Use --show-node-events instead to see node entry and exit events included in the timeline of failed and verbose specs.  Or you can run with -vv to always see all node events.  Lastly, --poll-progress-after and the PollProgressAfter decorator now provide a better mechanism for debugging specs that tend to get stuck. 
```

**Comment by furkatgofurov7** [comment]
@CecileRobertMichon @mboersma @Jont828 thanks folks for reviews and help! Most of the tests are green now, except `api-diff`. 
Based on https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2051 and fix landed in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2567 apidiff is testing only changes in the `api/` and `exp/api/` directories. I have not really touched them in this PR (minor doc ref fixes in /api folder only), can we consider this test as safe to ignore?

**Comment by CecileRobertMichon** [comment]
> I have not really touched them in this PR (minor doc ref fixes in /api folder only), can we consider this test as safe to ignore?

yes should be good

**Comment by CecileRobertMichon** [comment]
@furkatgofurov7 can you please remove the WIP from the title and hold label if it's ready for review/merge

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @mboersma @Jont828

**Comment by furkatgofurov7** [comment]
> /hold
> 
> I will separate Go version update changes from this PR to have a smoother transition overall 🤞🏼

/hold cancel

**Comment by furkatgofurov7** [comment]
> I have a fix for the conformance test, but I think that's just a deprecation warning, not an error:

@mboersma I have taken your suggested changes in earlier and it broke the conformance tests in `windows-containerd-upstream-with-ci-artifacts` [job](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2773/pull-cluster-api-provider-azure-windows-containerd-upstream-with-ci-artifacts/1598643537732177920) but have not looked much into it myself since it is a cosmetic change

File: Makefile
Comment (furkatgofurov7) [reply Makefile]: okay, found it in the CHANGELOG, it has been removed in v2.5.0 release of Ginkgo: https://github.com/onsi/ginkgo/blob/master/CHANGELOG.md#ginkgo-output-now-includes-a-timeline-view-of-the-spec

Comment (mboersma) [reply Makefile]: So it was deprecated in 2.0, removed in 2.5. Thanks for tracking this down, and sorry for my incomplete information.

File: Makefile
Comment (furkatgofurov7) [reply Makefile]: Yes, was busy with removing it after you confirmed, now should be fine.

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 this needs to be rebased, sorry :(

no problem, done

**Comment by mboersma** [comment]
> I am planning to take a look into it more closely in a follow-up PR if that makes sense.

That would be great, or just let's make sure we open an issue to track the idea of removing the `Measure` deprecation warning.

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Comment by jackfrancis** [comment]
Great job @furkatgofurov7 and @mboersma getting this over all the hurdles!

Are the apidiff complaints related to the comment changes?

**Comment by furkatgofurov7** [comment]
> Are the apidiff complaints related to the comment changes?

Thanks @jackfrancis! I believe so (best to my knowledge what it does and my findings), however please double check if that is not the case 🙂

**Comment by furkatgofurov7** [comment]
> That would be great, or just let's make sure we open an issue to track the idea of removing the `Measure` deprecation warning.

@mboersma here it is: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2903
I found out, another warning in the runs, so decided to include that as well in the same issue since they are both deprecation warnings at the end.

**Pull request merged**
